File: src/providers/openai-compatible.ts
/** One adapter for every OpenAI-compatible chat endpoint: OpenAI, DeepSeek, Groq, Together, OpenRouter, Ollama, Gemini. */
import type { CompletionRequest, CompletionResult, Provider } from "./types.js";
import { ProviderError } from "./types.js";

export interface OpenAICompatibleOptions {
  name: string;
  baseURL: string;
  apiKey?: string;
  /** OpenAI reasoning models reject `max_tokens` and need `max_completion_tokens`. Ollama is the reverse. Default "auto". */
  tokenParam?: "max_tokens" | "max_completion_tokens" | "auto";
  timeoutMs?: number;
  fetch?: typeof fetch;
}

export class OpenAICompatibleProvider implements Provider {
  readonly name: string;
  constructor(private opts: OpenAICompatibleOptions) { this.name = opts.name; }

  /** Build the JSON body and headers for one chat-completions request. Shared by complete() and streamRaw(). */
  private buildRequest(req: CompletionRequest, stream: boolean): { url: string; headers: Record<string, string>; body: Record<string, unknown> } {
    const url = `${this.opts.baseURL.replace(/\/$/, "")}/chat/completions`;
    const tokenParam = this.resolveTokenParam();
    const body: Record<string, unknown> = {
      model: req.model,
      messages: req.messages.map((m) => ({ role: m.role, content: m.content, ...(m.name ? { name: m.name } : {}) })),
      ...(req.maxTokens ? { [tokenParam]: req.maxTokens } : {}),
      ...(req.temperature !== undefined ? { temperature: req.temperature } : {}),
      ...(req.tools?.length ? { tools: req.tools.map((t) => ({ type: "function", function: { name: t.name, description: t.description, parameters: t.parameters ?? { type: "object", properties: {} } } })) } : {}),
      ...(req.params ?? {}),
      ...(stream ? { stream: true } : {}),
    };
    const headers: Record<string, string> = { "Content-Type": "application/json" };
    if (this.opts.apiKey) headers.Authorization = `Bearer ${this.opts.apiKey}`;
    return { url, headers, body };
  }

  /** POST with the per-attempt timeout and caller abort wired in. Network failures become ProviderError. */
  private async post(url: string, headers: Record<string, string>, body: unknown, signal: AbortSignal | undefined, timeoutMs: number): Promise<Response> {
    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), timeoutMs);
    signal?.addEventListener("abort", () => ctrl.abort(), { once: true });
    try {
      return await (this.opts.fetch ?? fetch)(url, { method: "POST", headers, body: JSON.stringify(body), signal: ctrl.signal });
    } catch (e) {
      throw new ProviderError(this.name, null, e instanceof Error ? e.message : String(e));
    } finally {
      clearTimeout(timer);
    }
  }

  async complete(req: CompletionRequest, signal?: AbortSignal): Promise<CompletionResult> {
    const { url, headers, body } = this.buildRequest(req, false);
    const t0 = performance.now();
    const res = await this.post(url, headers, body, signal, this.opts.timeoutMs ?? 120_000);
    const text = await res.text();
    let json: any;
    try { json = JSON.parse(text); } catch { json = { raw: text }; }
    if (!res.ok) throw new ProviderError(this.name, res.status, json?.error?.message ?? text.slice(0, 300), json);

    const choice = json.choices?.[0] ?? {};
    const msg = choice.message ?? {};
    return {
      text: typeof msg.content === "string" ? msg.content : "",
      toolCalls: (msg.tool_calls ?? []).map((c: any) => ({ id: c.id, name: c.function?.name, arguments: c.function?.arguments ?? "{}" })),
      finishReason: choice.finish_reason ?? "unknown",
      usage: { inputTokens: json.usage?.prompt_tokens ?? 0, outputTokens: json.usage?.completion_tokens ?? 0 },
      servedModel: json.model ?? req.model,
      latencyMs: Math.round(performance.now() - t0),
      raw: json,
    };
  }

  /**
   * Start a streaming chat completion and return the upstream Response untouched, so a caller can pipe
   * `res.body` (server-sent events) straight through. The response headers have arrived; the body has not.
   * A non-2xx status is thrown as ProviderError after reading the error body.
   * The timeout here covers only the time to first byte; the stream itself is bounded by the caller's signal.
   */
  async streamRaw(req: CompletionRequest, signal?: AbortSignal): Promise<Response> {
    const { url, headers, body } = this.buildRequest(req, true);
    const res = await this.post(url, headers, body, signal, this.opts.timeoutMs ?? 120_000);
    if (!res.ok) {
      const text = await res.text();
      let json: any;
      try { json = JSON.parse(text); } catch { json = { raw: text }; }
      throw new ProviderError(this.name, res.status, json?.error?.message ?? text.slice(0, 300), json);
    }
    return res;
  }

  private resolveTokenParam(): "max_tokens" | "max_completion_tokens" {
    if (this.opts.tokenParam && this.opts.tokenParam !== "auto") return this.opts.tokenParam;
    // Verified 2026-09-17: api.openai.com rejects max_tokens on gpt-5.x; Ollama accepts both but is 60x slower on max_completion_tokens.
    return /api\.openai\.com/.test(this.opts.baseURL) ? "max_completion_tokens" : "max_tokens";
  }
}

File: python/tiershift/providers/openai_compatible.py
"""One adapter for every OpenAI-compatible chat endpoint: OpenAI, DeepSeek, Groq, Together, OpenRouter, Ollama, Gemini."""

from __future__ import annotations

import re
import time
from typing import Any, Literal, Optional

import httpx

from ..types import ToolCall
from .base import CompletionRequest, CompletionResult, ProviderError

TokenParam = Literal["max_tokens", "max_completion_tokens", "auto"]


class OpenAICompatibleProvider:
    def __init__(self, name: str, base_url: str, api_key: Optional[str] = None, token_param: TokenParam = "auto", timeout_s: float = 120.0, client: Optional[httpx.Client] = None) -> None:
        self.name = name
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.token_param = token_param
        self.timeout_s = timeout_s
        self._client = client

    def _resolve_token_param(self) -> str:
        if self.token_param != "auto":
            return self.token_param
        # Verified 2026-09-17: api.openai.com rejects max_tokens on gpt-5.x; Ollama accepts both but is 60x slower on max_completion_tokens.
        return "max_completion_tokens" if re.search(r"api\.openai\.com", self.base_url) else "max_tokens"

    def build_body(self, req: CompletionRequest) -> dict[str, Any]:
        body: dict[str, Any] = {
            "model": req.model,
            "messages": [{"role": m["role"], "content": m.get("content", ""), **({"name": m["name"]} if m.get("name") else {})} for m in req.messages],
        }
        if req.max_tokens:
            body[self._resolve_token_param()] = req.max_tokens
        if req.temperature is not None:
            body["temperature"] = req.temperature
        if req.tools:
            body["tools"] = [{"type": "function", "function": {"name": t["name"], "description": t.get("description"), "parameters": t.get("parameters") or {"type": "object", "properties": {}}}} for t in req.tools]
        if req.params:
            body.update(req.params)
        return body

    def complete(self, req: CompletionRequest) -> CompletionResult:
        url = f"{self.base_url}/chat/completions"
        headers = {"Content-Type": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        t0 = time.perf_counter()
        try:
            client = self._client or httpx.Client(timeout=self.timeout_s)
            try:
                res = client.post(url, json=self.build_body(req), headers=headers)
            finally:
                if self._client is None:
                    client.close()
        except httpx.HTTPError as e:
            raise ProviderError(self.name, None, str(e)) from e
        try:
            data: Any = res.json()
        except ValueError:
            data = {"raw": res.text}
        if res.status_code >= 400:
            msg = (data.get("error") or {}).get("message") if isinstance(data, dict) and isinstance(data.get("error"), dict) else None
            raise ProviderError(self.name, res.status_code, msg or res.text[:300], data)
        choice = (data.get("choices") or [{}])[0]
        msg_obj = choice.get("message") or {}
        content = msg_obj.get("content")
        return CompletionResult(
            text=content if isinstance(content, str) else "",
            tool_calls=[ToolCall(id=c.get("id", ""), name=(c.get("function") or {}).get("name", ""), arguments=(c.get("function") or {}).get("arguments", "{}")) for c in (msg_obj.get("tool_calls") or [])],
            finish_reason=choice.get("finish_reason") or "unknown",
            input_tokens=int((data.get("usage") or {}).get("prompt_tokens") or 0),
            output_tokens=int((data.get("usage") or {}).get("completion_tokens") or 0),
            served_model=data.get("model") or req.model,
            latency_ms=round((time.perf_counter() - t0) * 1000),
            raw=data,
        )
